include/unifex/scheduler_concepts.hpp

```diff
@@ -25,48 +25,144 @@
 
 namespace unifex {
 namespace _schedule {
+  struct _fn;
+
+  template <typename Scheduler>
+  UNIFEX_CONCEPT_FRAGMENT( //
+    _with_tag_invoke_helper, //
+      requires () (0) &&
+      tag_invocable<_fn, Scheduler> &&
+      sender<tag_invoke_result_t<_fn, Scheduler>>);
+  template <typename Scheduler>
+  UNIFEX_CONCEPT //
+    _with_tag_invoke = //
+      UNIFEX_FRAGMENT(_schedule::_with_tag_invoke_helper, Scheduler);
+
+  template <typename Scheduler>
+  using _member_schedule_result_t =
+    decltype(UNIFEX_DECLVAL(Scheduler).schedule());
+
+  template <typename Scheduler>
+  UNIFEX_CONCEPT_FRAGMENT( //
+    _has_member_schedule_, //
+      requires() (0) &&
+      sender<_member_schedule_result_t<Scheduler>>);
+  template <typename Scheduler>
+  UNIFEX_CONCEPT //
+    _with_member_schedule = //
+      UNIFEX_FRAGMENT(_schedule::_has_member_schedule_, Scheduler);
+
+  // For adapting an executor as a sender.
+  template <typename Executor>
+  struct _as_sender {
+    struct type {
+    private:
+      Executor ex_;
+    public:
+      template <template <class...> class Variant,
+                template <class...> class Tuple>
+      using value_types = Variant<Tuple<>>;
+
+      template <template <class...> class Variant>
+      using error_types = Variant<std::exception_ptr>;
+
+      static constexpr bool sends_done = true;
+
+      explicit type(Executor e) noexcept
+        : ex_((Executor&&) e) {}
+
+      template(typename Self, typename Receiver)
+        (requires same_as<remove_cvref_t<Self>, type> AND receiver_of<Receiver>)
+      friend connect_result_t<member_t<Self, Executor>, Receiver>
+      tag_invoke(tag_t<connect>, Self&& self, Receiver&& r) {
+        return connect(((Self&&) self).ex_, (Receiver&&) r);
+      }
+    };
+  };
+  template <typename Executor>
+  using _as_sender_t = typename _as_sender<remove_cvref_t<Executor>>::type;
+
   struct sender;
   inline const struct _fn {
   private:
     template <typename Scheduler>
-    using _schedule_member_result_t =
-      decltype(UNIFEX_DECLVAL(Scheduler).schedule());
+    static auto _select() noexcept {
+      if constexpr (_with_tag_invoke<Scheduler>) {
+        return meta_tag_invoke_result<_fn>{};
+      } else if constexpr (_with_member_schedule<Scheduler>) {
+        return meta_quote1<_member_schedule_result_t>{};
+      } else if constexpr (executor<Scheduler>) {
+        return meta_quote1<_as_sender_t>{};
+      } else {
+        return type_always<void>{};
+      }
```

Comment:
Is the use of this _select() indirection here to avoid instantiating the return-types for `operator()` overloads that are not selected below?

Reply:
Yes.